sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ReceiverManager.cs

```diff
@@ -0,0 +1,358 @@
+                    // to finish
+                    await Receiver.CompleteAsync(
+                        message.LockToken,
+                        CancellationToken.None)
+                        .ConfigureAwait(false);
+                }
+
+                await CancelTask(renewLockCancellationTokenSource, renewLock).ConfigureAwait(false);
+            }
+            catch (Exception ex) when (!(ex is TaskCanceledException))
+            {
+                ThrowIfSessionLockLost(ex, errorSource);
+                await RaiseExceptionReceived(
+                    new ProcessErrorEventArgs(
+                        ex,
+                        errorSource,
+                        _fullyQualifiedNamespace,
+                        _entityPath))
+                    .ConfigureAwait(false);
+
+                // if the user settled the message, or if the message or session lock was lost,
+                // do not attempt to abandon the message
+                ServiceBusException.FailureReason? failureReason = (ex as ServiceBusException)?.Reason;
+                if (!message.IsSettled &&
+                    _receiverOptions.ReceiveMode == ReceiveMode.PeekLock &&
+                    failureReason != ServiceBusException.FailureReason.SessionLockLost &&
+                    failureReason != ServiceBusException.FailureReason.MessageLockLost)
+                {
+                    try
+                    {
+                        await Receiver.AbandonAsync(message.LockToken).ConfigureAwait(false);
+                    }
+                    catch (Exception exception)
+                    {
+                        ThrowIfSessionLockLost(exception, ServiceBusErrorSource.Abandon);
+                        await RaiseExceptionReceived(
+                            new ProcessErrorEventArgs(
+                                exception,
+                                ServiceBusErrorSource.Abandon,
+                                _fullyQualifiedNamespace,
+                                _entityPath))
+                        .ConfigureAwait(false);
+                    }
+                }
+            }
+            finally
+            {
+                renewLockCancellationTokenSource?.Cancel();
+                renewLockCancellationTokenSource?.Dispose();
+            }
+        }
+
+        public virtual async Task OnMessageHandler(ServiceBusReceivedMessage message, CancellationToken processorCancellationToken)
+        {
+            var args = new ProcessMessageEventArgs(
+                message,
+                Receiver,
+                processorCancellationToken);
+            await _messageHandler(args).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        ///
+        /// </summary>
+        /// <param name="message"></param>
+        /// <param name="cancellationTokenSource"></param>
+        /// <returns></returns>
+        private async Task RenewMessageLock(
+        ServiceBusReceivedMessage message,
```

Comment:
nit: indent

Reply:
This method is almost the copy of RenewSessionLock

Reply:
It is similar but there are enough differences that it would be worse I think to try to merge.